File: backend/services/rollback.py
"""Rollback service for undoing Meta campaign deployments."""
import logging
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from models.campaign import Project, ProjectStatus
from models.job import DeployLog, DeployStep, DeployStatus
from services.meta_api import MetaAPIService

logger = logging.getLogger(__name__)

# Rollback order: reverse of creation order
ROLLBACK_ORDER = [
    DeployStep.ad,
    DeployStep.ad_creative,
    DeployStep.adset,
    DeployStep.campaign,
]

STEP_TO_ENTITY_TYPE = {
    DeployStep.ad: "ad",
    DeployStep.ad_creative: "ad_creative",
    DeployStep.adset: "adset",
    DeployStep.campaign: "campaign",
}


async def rollback_project(
    db: AsyncSession,
    meta_service: MetaAPIService,
    project_id: int,
    keep_partial: bool = False,
) -> dict:
    """
    Rollback a deployed project by deleting all created Meta entities.

    Args:
        keep_partial: If True, only rollback failed entities. If False, rollback everything.

    Returns:
        {"rolled_back": [...], "errors": [...]}
    """
    result = await db.execute(
        select(DeployLog)
        .where(DeployLog.project_id == project_id)
        .where(DeployLog.meta_id.isnot(None))
        .order_by(DeployLog.id.desc())
    )
    logs = list(result.scalars().all())

    if keep_partial:
        # Only rollback entries that are marked as failed
        logs = [l for l in logs if l.status == DeployStatus.failed]

    rolled_back = []
    errors = []

    # Process in reverse order (ads first, then adsets, then campaign)
    for step in ROLLBACK_ORDER:
        step_logs = [l for l in logs if l.step == step and l.status != DeployStatus.rolled_back]
        for log_entry in step_logs:
            entity_type = STEP_TO_ENTITY_TYPE.get(log_entry.step)
            if not entity_type or not log_entry.meta_id:
                continue

            # Skip creative uploads - they don't need rollback
            if log_entry.step == DeployStep.creative_upload:
                continue

            try:
                await meta_service.delete_entity(entity_type, log_entry.meta_id)
                log_entry.status = DeployStatus.rolled_back
                rolled_back.append({
                    "type": entity_type,
                    "meta_id": log_entry.meta_id,
                    "name": log_entry.entity_name,
                })
                logger.info(f"Rolled back {entity_type} {log_entry.meta_id}")
            except Exception as e:
                error_msg = f"Failed to rollback {entity_type} {log_entry.meta_id}: {e}"
                errors.append(error_msg)
                logger.error(error_msg)

    # Update project status
    project_result = await db.execute(
        select(Project).where(Project.id == project_id)
    )
    project = project_result.scalar_one_or_none()
    if project:
        project.status = ProjectStatus.failed
        project.meta_campaign_id = None

    await db.commit()

    return {"rolled_back": rolled_back, "errors": errors}

File: api/lib/services/rollback.ts
/**
 * Rollback service for undoing Meta campaign deployments.
 */
import { eq, isNotNull, desc, and } from "drizzle-orm";
import type { Database } from "../db/client.js";
import { projects, deployLogs } from "../db/schema.js";
import { MetaApiService } from "./meta-api.js";

// ── Constants ───────────────────────────────────────────────────────────────

type DeployStep = "creative_upload" | "campaign" | "adset" | "ad_creative" | "ad";

// Rollback order: reverse of creation order
const ROLLBACK_ORDER: DeployStep[] = [
  "ad",
  "ad_creative",
  "adset",
  "campaign",
];

const STEP_TO_ENTITY_TYPE: Record<string, string> = {
  ad: "ad",
  ad_creative: "ad_creative",
  adset: "adset",
  campaign: "campaign",
};

// ── Types ───────────────────────────────────────────────────────────────────

interface RolledBackEntry {
  type: string;
  meta_id: string;
  name: string | null;
}

interface RollbackResult {
  rolled_back: RolledBackEntry[];
  errors: string[];
}

// ── Public API ──────────────────────────────────────────────────────────────

/**
 * Rollback a deployed project by deleting all created Meta entities.
 *
 * @param keepPartial - If true, only rollback failed entities. If false, rollback everything.
 */
export async function rollbackProject(
  db: Database,
  metaService: MetaApiService,
  projectId: number,
  keepPartial: boolean = false,
): Promise<RollbackResult> {
  // Fetch all deploy logs with a meta_id, ordered newest first
  let logs = await db
    .select()
    .from(deployLogs)
    .where(
      and(
        eq(deployLogs.projectId, projectId),
        isNotNull(deployLogs.metaId),
      ),
    )
    .orderBy(desc(deployLogs.id));

  if (keepPartial) {
    // Only rollback entries that are marked as failed
    logs = logs.filter((l) => l.status === "failed");
  }

  const rolledBack: RolledBackEntry[] = [];
  const errors: string[] = [];

  // Process in reverse order (ads first, then adsets, then campaign)
  for (const step of ROLLBACK_ORDER) {
    const stepLogs = logs.filter(
      (l) => l.step === step && l.status !== "rolled_back",
    );

    for (const logEntry of stepLogs) {
      const entityType = STEP_TO_ENTITY_TYPE[logEntry.step];
      if (!entityType || !logEntry.metaId) {
        continue;
      }

      // Skip creative uploads - they don't need rollback
      if (logEntry.step === "creative_upload") {
        continue;
      }

      try {
        await metaService.deleteEntity(entityType, logEntry.metaId);

        // Mark as rolled back
        await db
          .update(deployLogs)
          .set({ status: "rolled_back" })
          .where(eq(deployLogs.id, logEntry.id));

        rolledBack.push({
          type: entityType,
          meta_id: logEntry.metaId,
          name: logEntry.entityName,
        });

        console.log(`Rolled back ${entityType} ${logEntry.metaId}`);
      } catch (e) {
        const errorMsg = `Failed to rollback ${entityType} ${logEntry.metaId}: ${e}`;
        errors.push(errorMsg);
        console.error(errorMsg);
      }
    }
  }

  // Update project status
  const [project] = await db
    .select()
    .from(projects)
    .where(eq(projects.id, projectId));

  if (project) {
    await db
      .update(projects)
      .set({ status: "failed", metaCampaignId: null })
      .where(eq(projects.id, projectId));
  }

  return { rolled_back: rolledBack, errors };
}
